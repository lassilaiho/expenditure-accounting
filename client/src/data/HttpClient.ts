export class ConnectionError extends Error {
  public readonly name = 'ConnectionError';

  public get message() {
    return this.cause.message;
  }

  public constructor(public readonly cause: Error, ...params: any[]) {
    super(...params);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ConnectionError);
    }
  }
}

async function apiFetch(input: RequestInfo, init?: RequestInit | undefined) {
  try {
    return await fetch(input, init);
  } catch (e) {
    if (e instanceof Error) {
      throw new ConnectionError(e);
    }
    throw e;
  }
}

export default class HttpClient {
  public baseUrl = '';

  public defaultHeaders: Record<string, string> = {
    Authorization: '',
  };

  private _sessionToken: string | null = null;
  public get sessionToken() {
    return this._sessionToken;
  }
  public set sessionToken(token: string | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
