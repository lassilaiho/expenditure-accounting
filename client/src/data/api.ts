export type LoadState =
  | 'not-started'
  | 'loading'
  | 'finished'
  | 'failed';

export class SessionToken {
  public constructor(
    public readonly token: string,
    public readonly expiryTime: Date,
  ) { }

  public static fromJson(json: any): SessionToken | null {
    if (typeof json.token !== 'string'
      || typeof json.expiryTime !== 'string') {
      throw new Error('Invalid json object');
    }
    return new SessionToken(json.token, new Date(json.expiryTime));
  }

  public get isValid() {
    return this.expiryTime.getTime() > Date.now();
  }

  public toJSON(): unknown {
    return {
      token: this.token,
      expiryTime: this.expiryTime.toUTCString(),
    };
  }
}

export default class Api {
  public baseUrl: string = '';

  public defaultHeaders: Record<string, string> = {
    'Authorization': ''
  };

  private _sessionToken: SessionToken | null = null;
  public get sessionToken() { return this._sessionToken; }
  public set sessionToken(token: SessionToken | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token.token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
