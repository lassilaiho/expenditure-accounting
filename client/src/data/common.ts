import { createAction } from '@reduxjs/toolkit';

import { Product } from './products';
import { Purchase } from './purchases';
import { Tag } from './tags';

export type Collection<T> = {
  byId: { [id: number]: T };
  all: number[];
};

export function emptyCollection<T>(): Collection<T> {
  return { byId: {}, all: [] };
}

export const id = <T>(x: T) => x;

export const clearRemoteData = createAction('clearRemoteData');
export type RemoteDataSet = [Purchase, Product, Tag[]][];
export const setRemoteData = createAction<RemoteDataSet>('setRemoteData');
