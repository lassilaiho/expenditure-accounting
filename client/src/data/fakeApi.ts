import Api from './api';

export type FakeResponse = (
  method: string,
  url: string,
  body?: any,
) => Promise<Response>;

export class FakeApi extends Api {
  public constructor(private fakeResponse: FakeResponse) {
    super('');
  }

  public async get(url: string) {
    return this.fakeResponse('get', url);
  }

  public async postJson(url: string, body: any) {
    return this.fakeResponse('post', url, body);
  }

  public async patchJson(url: string, body: any) {
    return this.fakeResponse('patch', url, body);
  }

  public async delete(url: string) {
    return this.fakeResponse('delete', url);
  }
}

function jsonResponse(body: any) {
  return new Response(JSON.stringify(body), {
    headers: {
      'Content-Type': 'application/json',
    },
    status: 200,
  });
}

const tags = [
  {
    id: 1,
    name: 'tag 1',
  },
  {
    id: 2,
    name: 'tag 2',
  },
  {
    id: 3,
    name: 'tag 3',
  },
];
const products = [
  {
    id: 1,
    name: 'product 1',
  },
  {
    id: 2,
    name: 'product 2',
  },
];
const purchases = [
  {
    id: 1,
    product: products[0],
    date: '2020-03-14',
    quantity: '1.0',
    price: '2.0',
    tags: [tags[0], tags[2]],
  },
  {
    id: 2,
    product: products[0],
    date: '2020-01-14',
    quantity: '10',
    price: '13.2',
    tags: [tags[2]],
  },
  {
    id: 3,
    product: products[1],
    date: '2020-03-14',
    quantity: '0.8',
    price: '0.2',
    tags: [tags[0], tags[1], tags[2]],
  },
];

export const fakeData = { tags, products, purchases };

export const fakeApi = new FakeApi(async (method, url, body) => {
  switch (method + url) {
    case 'get/purchases':
      return jsonResponse({ purchases });
  }
  return new Response(null, { status: 404 });
});
