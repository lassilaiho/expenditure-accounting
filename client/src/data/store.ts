import Big from 'big.js';
import {
  action,
  computed,
  makeObservable,
  observable,
  runInAction,
} from 'mobx';
import moment from 'moment';
import React, { useContext } from 'react';

import { ensureOk } from '../util';
import Api, { LoadState } from './api';
import * as jsonConv from './jsonConvert';

export class Tag {
  public id: number;
  public name: string;

  public constructor(id: number, name: string) {
    makeObservable(this, {
      id: observable,
      name: observable,
    });
    this.id = id;
    this.name = name;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Tag(jsonConv.toNumber(json.id), jsonConv.toString(json.name));
  }

  public lowerCaseMatch(s: string) {
    return this.name.toLocaleLowerCase().includes(s);
  }
}

export class Purchase {
  public id: number;
  public product: Product;
  public date: moment.Moment;
  public quantity: Big;
  public price: Big;
  public get totalPrice() {
    return this.quantity.mul(this.price);
  }
  public tags: Tag[];
  public get tagsSortedByName() {
    return this.tags.slice().sort((a, b) => {
      const aName = a.name.toLocaleLowerCase();
      const bName = b.name.toLocaleLowerCase();
      if (aName < bName) {
        return -1;
      }
      if (aName > bName) {
        return 1;
      }
      return 0;
    });
  }

  public constructor(
    id: number,
    product: Product,
    date: moment.Moment,
    quantity: Big,
    price: Big,
    tags: Tag[],
  ) {
    makeObservable(this, {
      id: observable,
      product: observable,
      date: observable.ref,
      quantity: observable.ref,
      price: observable.ref,
      totalPrice: computed,
      tags: observable,
      tagsSortedByName: computed,
    });
    this.id = id;
    this.product = product;
    this.date = date;
    this.quantity = quantity;
    this.price = price;
    this.tags = tags;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Purchase(
      jsonConv.toNumber(json.id),
      Product.fromJson(json.product),
      jsonConv.toMomentUtc(json.date),
      jsonConv.toBig(json.quantity),
      jsonConv.toBig(json.price),
      jsonConv.toArray(Tag.fromJson, json.tags),
    );
  }

  public lowerCaseMatch(s: string) {
    return (
      this.product.lowerCaseMatch(s) || this.tags.some(t => t.lowerCaseMatch(s))
    );
  }
}

export class Product {
  public id: number;
  public name: string;

  public constructor(id: number, name: string) {
    makeObservable(this, {
      id: observable,
      name: observable,
    });
    this.id = id;
    this.name = name;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Product(
      jsonConv.toNumber(json.id),
      jsonConv.toString(json.name),
    );
  }

  public lowerCaseMatch(s: string) {
    return this.name.toLocaleLowerCase().includes(s);
  }
}

export class Store {
  public dataState: LoadState = 'not-started';
  public purchases: Purchase[] = [];
  public get purchasesById() {
    const map = new Map<number, Purchase>();
    for (const p of this.purchases) {
      map.set(p.id, p);
    }
    return map;
  }
  public tagsById = new Map<number, Tag>();
  public get tags() {
    const tags: Tag[] = [];
    for (const tag of this.tagsById.values()) {
      tags.push(tag);
    }
    return tags;
  }
  private get tagsByName() {
    const map = new Map<string, Tag>();
    for (const tag of this.tagsById.values()) {
      map.set(tag.name.toLowerCase(), tag);
    }
    return map;
  }
  public productsById = new Map<number, Product>();
  public get products() {
    const products: Product[] = [];
    for (const product of this.productsById.values()) {
      products.push(product);
    }
    return products;
  }
  private get productsByName() {
    const map = new Map<string, Product>();
    for (const product of this.productsById.values()) {
      map.set(product.name.toLowerCase(), product);
    }
    return map;
  }

  public constructor(private api: Api) {
    makeObservable<
      Store,
      | 'tagsByName'
      | 'productsByName'
      | 'parsePurchaseJson'
      | 'parsePurchasesJson'
    >(this, {
      dataState: observable,
      purchases: observable,
      purchasesById: computed,
      tagsById: observable,
      tags: computed,
      tagsByName: computed,
      productsById: observable,
      products: computed,
      productsByName: computed,
      addTags: action,
      reloadData: action,
      parsePurchasesJson: action,
      parsePurchaseJson: action,
      restorePurchase: action,
    });
  }

  public async updatePurchase(id: number) {
    const purchase = this.purchasesById.get(id);
    if (purchase) {
      ensureOk(
        await this.api.patchJson(`/purchases/${purchase.id}`, {
          product: purchase.product.id,
          date: purchase.date.format(),
          price: purchase.price,
          quantity: purchase.quantity,
          tags: purchase.tags.map(t => t.id),
        }),
      );
    } else {
      throw new Error(`No purchase with id: ${id}`);
    }
  }

  public async addTags(names: string[]) {
    const newTags: string[] = [];
    const result: Tag[] = [];
    for (const name of names) {
      const tag = this.getTagByName(name);
      if (tag === undefined) {
        newTags.push(name);
      } else {
        result.push(tag);
      }
    }
    if (newTags.length > 0) {
      const resp = await this.api.postJson('/tags', { tags: newTags });
      ensureOk(resp);
      const respJson = await resp.json();
      runInAction(() => {
        for (const tagJson of respJson.tags) {
          const tag = Tag.fromJson(tagJson);
          this.tagsById.set(tag.id, tag);
          result.push(tag);
        }
      });
    }
    return result;
  }

  public async addProduct(name: string) {
    const product = this.getProductByName(name);
    if (product === undefined) {
      const resp = await this.api.postJson('/products', { name });
      ensureOk(resp);
      const product = Product.fromJson(await resp.json());
      runInAction(() => this.productsById.set(product.id, product));
      return product;
    }
    return product;
  }

  public async addPurchase(purchase: Purchase) {
    const resp = await this.api.postJson('/purchases', {
      product: purchase.product.id,
      date: purchase.date.format(),
      price: purchase.price,
      quantity: purchase.quantity,
      tags: purchase.tags.map(t => t.id),
    });
    ensureOk(resp);
    const respJson = jsonConv.toObject(await resp.json());
    const id = jsonConv.toNumber(respJson.id);
    runInAction(() => {
      purchase.id = id;
      for (let i = 0; i < this.purchases.length; i++) {
        if (purchase.date > this.purchases[i].date) {
          this.purchases.splice(i, 0, purchase);
          return;
        }
      }
      this.purchases.splice(this.purchases.length, 0, purchase);
    });
  }

  public async deletePurchase(id: number) {
    const i = this.purchases.findIndex(p => p.id === id);
    if (i < 0) {
      throw new Error(`Purchase with id ${id} does not exist`);
    }
    ensureOk(await this.api.delete(`/purchases/${id}`));
    runInAction(() => this.purchases.splice(i, 1));
  }

  public async restorePurchase(id: number) {
    const resp = await this.api.postJson(`/purchases/${id}/restore`, {});
    await ensureOk(resp);
    const purchase = this.parsePurchaseJson((await resp.json()).purchase);
    runInAction(() => {
      for (let i = 0; i < this.purchases.length; i++) {
        if (purchase.date > this.purchases[i].date) {
          this.purchases.splice(i, 0, purchase);
          return;
        }
      }
      this.purchases.splice(this.purchases.length, 0, purchase);
    });
  }

  public async getLatestPurchaseByProduct(productName: string) {
    productName = productName.toLocaleLowerCase();
    return (
      this.purchases.find(p => p.product.lowerCaseMatch(productName)) ?? null
    );
  }

  private getTagByName(name: string) {
    return this.tagsByName.get(name.toLowerCase());
  }

  private getProductByName(name: string) {
    return this.productsByName.get(name.toLowerCase());
  }

  public async reloadData() {
    try {
      this.dataState = 'loading';
      this.purchases.length = 0;
      this.tagsById.clear();
      this.productsById.clear();

      const resp = await this.api.get('/purchases');
      ensureOk(resp);

      const json = await resp.json();
      runInAction(() => {
        this.parsePurchasesJson(json);
        this.dataState = 'finished';
      });
    } catch (e) {
      runInAction(() => (this.dataState = 'failed'));
      throw e;
    }
  }

  private parsePurchasesJson(data: any) {
    const purchases = data.purchases;
    if (!Array.isArray(purchases)) {
      throw new Error('Invalid response json');
    }
    for (const obj of purchases) {
      this.purchases.push(this.parsePurchaseJson(obj));
    }
  }

  private parsePurchaseJson(data: any) {
    const purchase = Purchase.fromJson(data);
    const product = this.productsById.get(purchase.product.id);
    if (product) {
      purchase.product = product;
    } else {
      this.productsById.set(purchase.product.id, purchase.product);
    }
    for (let i = 0; i < purchase.tags.length; i++) {
      const currentTag = purchase.tags[i];
      const existingTag = this.tagsById.get(currentTag.id);
      if (existingTag) {
        purchase.tags[i] = existingTag;
      } else {
        this.tagsById.set(currentTag.id, currentTag);
      }
    }
    return purchase;
  }
}

export const StoreContext = React.createContext<Store | null>(null);

export function useStore() {
  const store = useContext(StoreContext);
  if (store === null) {
    throw new Error('Store must not be null');
  }
  if (store.dataState === 'not-started') {
    store.reloadData();
  }
  return store;
}
