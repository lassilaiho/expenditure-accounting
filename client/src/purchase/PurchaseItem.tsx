import {
  Box,
  Button,
  Chip,
  Divider,
  Accordion,
  AccordionActions,
  AccordionDetails,
  AccordionSummary,
  Typography,
} from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import React from 'react';

import {
  getProductById,
  getTagsSortedByName,
  Purchase,
  totalPrice,
  useAppSelector,
} from '../data/store';
import { currency, numOfBig, threeDecimals } from '../util';

export interface PurchaseListItemProps {
  purchase: Purchase;
  expanded: boolean;
  onToggle: () => void;
  onEdit: (purchase: Purchase) => void;
  onDelete: (purchase: Purchase) => void;
}

const PurchaseItem: React.FC<PurchaseListItemProps> = props => {
  const { purchase: p, expanded, onToggle } = props;
  const product = useAppSelector(getProductById(p.product));
  const tags = useAppSelector(getTagsSortedByName(p.tags));

  let quantity = '';
  if (!p.quantity.eq(1)) {
    if (p.quantity.round(0).eq(p.quantity)) {
      quantity = `${p.quantity.toFixed(0)} × `;
    } else {
      quantity = `${threeDecimals.format(numOfBig(p.quantity))} × `;
    }
  }

  function onEdit(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    e.stopPropagation();
    props.onEdit(p);
  }

  function onDelete(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    e.stopPropagation();
    props.onDelete(p);
  }

  const formattedPrice = currency.format(numOfBig(totalPrice(p)));

  return (
    <Accordion
      expanded={expanded}
      onClick={onToggle}
      TransitionProps={{ unmountOnExit: true }}
    >
      <AccordionSummary expandIcon={<ExpandMoreIcon />}>
        <Box>
          <Typography>{quantity + product.name}</Typography>
          <Typography color='textSecondary'>
            {`${p.date.format('D.M.YYYY')} • ${formattedPrice}`}
          </Typography>
        </Box>
      </AccordionSummary>
      <AccordionDetails>
        <Box display='flex' flexWrap='wrap'>
          {tags.map(t => (
            <Box m={1} key={t.id}>
              <Chip label={t.name} />
            </Box>
          ))}
        </Box>
      </AccordionDetails>
      <Divider />
      <AccordionActions>
        <Button size='small' color='primary' onClick={onDelete}>
          Delete
        </Button>
        <Button size='small' color='primary' onClick={onEdit}>
          Edit
        </Button>
      </AccordionActions>
    </Accordion>
  );
};

export default PurchaseItem;
