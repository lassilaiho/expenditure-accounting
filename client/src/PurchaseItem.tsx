import {
  Box,
  Button,
  Chip,
  Divider,
  ExpansionPanel,
  ExpansionPanelActions,
  ExpansionPanelDetails,
  ExpansionPanelSummary,
  Typography
} from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import React from 'react';

import { Purchase } from './data/store';
import { currency, threeDecimals } from './util';

export interface PurchaseListItemProps {
  purchase: Purchase;
  expanded: boolean;
  onToggle: () => void;
  onEdit: (purchase: Purchase) => void;
  onDelete: (purchase: Purchase) => void;
}

const PurchaseItem: React.FC<PurchaseListItemProps> = props => {
  const { purchase: p, expanded, onToggle } = props;
  let quantity = '';
  if (p.quantity !== 1) {
    if (Number.isInteger(p.quantity)) {
      quantity = `${p.quantity.toFixed(0)} × `;
    } else {
      quantity = `${threeDecimals.format(p.quantity)} × `;
    }
  }

  function onEdit(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    e.stopPropagation();
    props.onEdit(p);
  }

  function onDelete(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    e.stopPropagation();
    props.onDelete(p);
  }

  return (
    <ExpansionPanel
      expanded={expanded}
      onClick={onToggle}
      TransitionProps={{ unmountOnExit: true }}
    >
      <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
        <Box>
          <Typography>{quantity + p.product.name}</Typography>
          <Typography color='textSecondary'>
            {`${p.date.format('D.M.YYYY')} • ${currency.format(p.totalPrice)}`}
          </Typography>
        </Box>
      </ExpansionPanelSummary>
      <ExpansionPanelDetails>
        <Box display='flex' flexWrap='wrap'>
          {p.tagsSortedByName.map(t => (
            <Box m={1} key={t.id}>
              <Chip label={t.name} />
            </Box>
          ))}
        </Box>
      </ExpansionPanelDetails>
      <Divider />
      <ExpansionPanelActions>
        <Button size='small' color='primary' onClick={onDelete}>
          Delete
        </Button>
        <Button size='small' color='primary' onClick={onEdit}>
          Edit
        </Button>
      </ExpansionPanelActions>
    </ExpansionPanel>
  );
};

export default PurchaseItem;
