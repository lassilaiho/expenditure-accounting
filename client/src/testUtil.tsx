import MomentUtil from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { render } from '@testing-library/react';
import React, { useRef } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { fakeApi } from './data/fakeApi';
import { Store, StoreContext } from './data/store';

const TestProviders = (defaultStore?: Store, history?: string[]) => (
  props: any,
) => {
  const store = useRef<Store | null>(null);
  if (store.current === null) {
    if (defaultStore === undefined) {
      store.current = new Store(fakeApi);
    } else {
      store.current = defaultStore;
    }
  }
  return (
    <StoreContext.Provider value={store.current}>
      <MuiPickersUtilsProvider utils={MomentUtil}>
        <MemoryRouter initialEntries={history}>{props.children}</MemoryRouter>
      </MuiPickersUtilsProvider>
    </StoreContext.Provider>
  );
};

export interface CustomRenderOptions {
  store?: Store;
  history?: string[];
}

const customRender = (ui: any, options?: CustomRenderOptions) => {
  let store, history;
  if (options) {
    store = options.store;
    history = options.history;
  }
  return render(ui, { wrapper: TestProviders(store, history) });
};

export * from '@testing-library/react';

export { customRender as render };
