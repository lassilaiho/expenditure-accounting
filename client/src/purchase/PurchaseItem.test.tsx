import React from 'react';

import { fakeApi } from '../data/fakeApi';
import { Purchase, Store } from '../data/store';
import { render } from '../testUtil';
import PurchaseItem from './PurchaseItem';

let store: Store;
let quantity1: Purchase, quantityIntNot1: Purchase, quantityDecNot1: Purchase;

beforeAll(async () => {
  store = new Store(fakeApi);
  await store.reloadData();
  quantity1 = store.purchases[0];
  quantityIntNot1 = store.purchases[1];
  quantityDecNot1 = store.purchases[2];
});

test('renders correctly with purchase with quantity 1', () => {
  const { asFragment } = render(
    <PurchaseItem
      purchase={quantity1}
      expanded={false}
      onToggle={() => {}}
      onEdit={() => {}}
      onDelete={() => {}}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with purchase with integer quantity not 1', async () => {
  const { asFragment } = render(
    <PurchaseItem
      purchase={quantityIntNot1}
      expanded={false}
      onToggle={() => {}}
      onEdit={() => {}}
      onDelete={() => {}}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with purchase with decimal quantity not 1', async () => {
  const { asFragment } = render(
    <PurchaseItem
      purchase={quantityDecNot1}
      expanded={false}
      onToggle={() => {}}
      onEdit={() => {}}
      onDelete={() => {}}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly when expanded', async () => {
  const { asFragment } = render(
    <PurchaseItem
      purchase={quantityDecNot1}
      expanded={true}
      onToggle={() => {}}
      onEdit={() => {}}
      onDelete={() => {}}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});
