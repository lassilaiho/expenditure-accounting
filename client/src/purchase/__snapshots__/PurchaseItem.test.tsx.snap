// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when expanded 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="true"
      class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content Mui-expanded"
      >
        <div
          class="MuiBox-root MuiBox-root-4"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            0,800 × product 2
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            14.3.2020 • 0,16 €
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-entered"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
            >
              <div
                class="MuiBox-root MuiBox-root-5"
              >
                <div
                  class="MuiBox-root MuiBox-root-6"
                >
                  <div
                    class="MuiChip-root"
                  >
                    <span
                      class="MuiChip-label"
                    >
                      tag 1
                    </span>
                  </div>
                </div>
                <div
                  class="MuiBox-root MuiBox-root-7"
                >
                  <div
                    class="MuiChip-root"
                  >
                    <span
                      class="MuiChip-label"
                    >
                      tag 2
                    </span>
                  </div>
                </div>
                <div
                  class="MuiBox-root MuiBox-root-8"
                >
                  <div
                    class="MuiChip-root"
                  >
                    <span
                      class="MuiChip-label"
                    >
                      tag 3
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <hr
              class="MuiDivider-root"
            />
            <div
              class="MuiAccordionActions-root MuiAccordionActions-spacing"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Delete
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Edit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with purchase with decimal quantity not 1 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        <div
          class="MuiBox-root MuiBox-root-3"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            0,800 × product 2
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            14.3.2020 • 0,16 €
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with purchase with integer quantity not 1 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            10 × product 1
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            14.1.2020 • 132,00 €
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly with purchase with quantity 1 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        <div
          class="MuiBox-root MuiBox-root-1"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            product 1
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            14.3.2020 • 2,00 €
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
